### PR TITLE
Implement Ultra Burst

### LINKED ISSUE #2288: Implement ULTRA BURST
T'is missing :eyes:

https://bulbapedia.bulbagarden.net/wiki/Ultra_Burst

### PR DESCRIPTION
**Pull request by kittenchilly**
## Description
Adds the ability for Necrozma Dusk Mane and Necrozma Dawn Wings to undergo Ultra Burst if holding an Ultranecrozium Z and the player has a Z Ring, transforming them into Ultra Necrozma.
Ultra Burst can be defined by putting the relevant pokemon in a form table with a z crystal held item, much like mega evolution. It is triggered and used almost exactly like mega evolution, except instead of a mega stone and mega ring, its a z crystal and z ring.

## Images
ultraburst2

## Issue(s) that this PR fixes
Closes #2288

## **Discord contact info**
kittenchilly

### CODE REVIEW

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -8603,13 +8605,15 @@ static void HandleScriptMegaPrimal(u32 caseId, u32 battlerId, bool32 isMega)
     // Change species.
     if (caseId == 0)
     {
-        if (isMega)
+        if (type == 0)
```
Comment (LOuroboros) [inline src/battle_script_commands.c:8608]: Can you add some constant labels for the `type` argument?

That'd make the code more readable :eyes:

Comment (kittenchilly) [reply src/battle_script_commands.c:8608]: Oh yeah, that was something I was going to get to but I guess I forgot

**Comment by AsparagusEduardo** [comment]
Just like how Mega Evolution has tests, Ultra Burst needs tests as well.
I recommend the following, based on Mega Evolution's:
- "Dusk Mane Necrozma can Ultra Burst holding Ultranecrozium Z"
- "Ultra Burst's order is determined by Speed - opponent faster"
- "Ultra Burst's order is determined by Speed - player faster"
- "Ultra Burst affects turn order"
- "Ultra Burst happens after switching, but before Focus Punch-like Moves"
- "Ultra Burst and Mega Evolution can happen on the same turn"

**Comment by kittenchilly** [comment]
> Just like how Mega Evolution has tests, Ultra Burst needs tests as well. I recommend the following, based on Mega Evolution's:
> 
> * "Dusk Mane Necrozma can Ultra Burst holding Ultranecrozium Z"
> * "Ultra Burst's order is determined by Speed - opponent faster"
> * "Ultra Burst's order is determined by Speed - player faster"
> * "Ultra Burst affects turn order"
> * "Ultra Burst happens after switching, but before Focus Punch-like Moves"
> * "Ultra Burst and Mega Evolution can happen on the same turn"

Alright, done, and they all pass.

**Comment by AsparagusEduardo** [comment]
Looks good, just needs to solve conflicts.

File: src/battle_controller_player_partner.c [hunk 2]
```diff
@@ -373,22 +373,14 @@ static void PlayerPartnerHandleChooseMove(u32 battler)
         // If partner can mega evolve, do it.
         if (CanMegaEvolve(battler))
             BtlController_EmitTwoReturnValues(BUFFER_B, 10, (chosenMoveId) | (RET_MEGA_EVOLUTION) | (gBattlerTarget << 8));
+		else if (CanUltraBurst(battler))
+			BtlController_EmitTwoReturnValues(BUFFER_B, 10, (chosenMoveId) | (RET_ULTRA_BURST) | (gBattlerTarget << 8));
```
Comment (AsparagusEduardo) [inline src/battle_controller_player_partner.c:377]: Spaces instead of tabs

**Comment by 97mos** [comment]
What about the indicator that replaces the Lv like for mega and the primal reversions?

**Comment by LOuroboros** [comment]
> What about the indicator that replaces the Lv like for mega and the primal reversions?

Ultra Burst doesn't have one of its own, Necrozma's level is displayed normally without any extra icons.

citra-qt_20230901_221414889

**Pull request merged**